### PR TITLE
[Dual-stack] Fix generateAPIPodStatus() of kubelet handling Secondary IP

### PR DESCRIPTION
**Pull request by lzhecheng**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
[Dual-stack] Fix generateAPIPodStatus() of kubelet handling Secondary IP. hostIPs order may not be be consistent. If secondary IP is before primary one, current logic adds primary IP twice into PodIPs, which leads to error: "may specify no more than one IP for each IP family".
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
[Dual-stack] Fix generateAPIPodStatus() of kubelet handling Secondary IP. hostIPs order may not be be consistent. If secondary IP is before primary one, current logic adds primary IP twice into PodIPs, which leads to error: "may specify no more than one IP for each IP family".
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go
Comment (aojea) [reply pkg/kubelet/kubelet_pods.go:1638]: I see, if the first hostIP returns IPA, and the subsequent request returms IPB,IPA in different order, it will add { IPA, IPA}

However, there are a lot of places in the code that depend on the order of the returned address, that is why we have
```
func (kl *Kubelet) sortPodIPs(podIPs []string) []string {
```

that is called from 
```
convertStatusToAPIStatus
```

/cc @danwinship he has a keep to try to organize this

Comment (jackfrancis) [reply pkg/kubelet/kubelet_pods.go:1638]: I also don't understand why the Node.Status.Addresses (which is the underlying slice data structure that `hostIPs` represents) would inconsistently represent a well-ordered set.

Comment (aojea) [reply pkg/kubelet/kubelet_pods.go:1638]: I don't remember if we had something about the order, typically this may be added by an external controller, but in this case it seems that the external controller is adding the addresses in reverse order

Comment (jackfrancis) [reply pkg/kubelet/kubelet_pods.go:1638]: If the problem is the order, then doing this additive mutation should really be transformed to just look for net new IPs, rather than assuming that the particular pathology is restricted to swapping the order of ipv4 and ipv6 IPs _in particular_. Does that make sense?

I think a solution like this is more general, and preferred:

```
$ git diff pkg/kubelet/kubelet_pods.go
diff --git a/pkg/kubelet/kubelet_pods.go b/pkg/kubelet/kubelet_pods.go
index fbc9104c428..85434f823b8 100644
--- a/pkg/kubelet/kubelet_pods.go
+++ b/pkg/kubelet/kubelet_pods.go
@@ -1635,7 +1635,11 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
                                }
                                // Secondary IP is not set #105320
                                if len(hostIPs) == 2 && len(s.PodIPs) == 1 {
-                                       s.PodIPs = append(s.PodIPs, v1.PodIP{IP: hostIPs[1].String()})
+                                       for _, ip := range hostIPs {
+                                               if ip.String() != s.PodIP {
+                                                       s.PodIPs = append(s.PodIPs, v1.PodIP{IP: ip.String()})
+                                               }
+                                       }
                                }
                        }
                }
```

This passes your new UT @lzhecheng 

cc @CecileRobertMichon who is also tracking this IPV6 + OOT cloud-provider-azure pathology

Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go:1638]: @aojea hello, do you suggest using `sortPodIPs()` for `hostIPs`? I think it may also work.
Yes, IPs are added by external controller, but `kl.getHostIPsAnyWay()` doesn't ensure the order right?

Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go:1638]: @jackfrancis I also thought about your suggested solution. But I think the key is: In PodIPs, IPs of same IP family should be only one. What do you think?

Comment (jackfrancis) [reply pkg/kubelet/kubelet_pods.go:1638]: If indeed there are strict combinatoric requirements (only ipv4 or ipv6 or ipv4+ipv6 but not ipv4+ipv4 or ipv6+ipv6 or > 2 IPs) then IMO that validation should be discrete and separate from this “add a secondary IP to an existing IP” flow. Dealing with those combinatoric requirements *and* dealing with reordering/merging in this single flow makes the code almost unreadable, it seems to me.

My 2¢

Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go:1638]: There's indeed such requirement, which is why the OOT + IPv6 issue comes.
I think reordering/merging is fine while picking a value from a fixed index seems dangerous. It's my opinion and let's have more people taking a look.

Comment (jackfrancis) [reply pkg/kubelet/kubelet_pods.go:1638]: sgtm

Comment (jackfrancis) [reply pkg/kubelet/kubelet_pods.go:1638]: I would also vote for that approach. Is someone working on that?

Comment (aojea) [reply pkg/kubelet/kubelet_pods.go:1638]: @lzhecheng ? 😄

Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go:1638]: I'm happy to solve this issue with the proper approach. But please correct me if I'm wrong.
The order here involves the hostIPs set last time, which can be got with `s.PodIPs`. So `getHostIPsAnyWay()` needs something like last time IP family as an input. Even if we do it this way, we cannot simply add the second IP into `PodIPs` because that will leads to the problem I met now (`<IPv6>, <getHostIPsAnyWay>`), so the rest of code still needs to check IP family.
There's another approach which avoids rest of code dealing with these things, which is overwriting PodIPs with host IPs. Can we do it? I'm not sure since I saw your comment mentioning "immutable once set". I don't know if overwriting will lead to other problems.

Comment (aojea) [reply pkg/kubelet/kubelet_pods.go:1638]: let me try to find some time this week to get to this, this part is very brittle and delicate and I'd like to be completely sure before I suggest anything

Comment (aojea) [reply pkg/kubelet/kubelet_pods.go:1638]: > I still feel like #106715 was wrong, and the right fix is "wait for your network to settle before you start kubelet".

to be more specific, it should be "wait for your cloud-controller-manager to write the node.status.Addresses and in the right order?"

Comment (aojea) [reply pkg/kubelet/kubelet_pods.go:1638]: @danwinship we have some expectations on the order and nodeIP is the source of truth

https://github.com/kubernetes/kubernetes/pull/97979 

> kubelet is assuming that the IPs are returned in preference order

In this case the problem is the controller is returning an order that doesn't match what we defined as the primary IP family of the kubelet.

Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go:1638]: Ok, the order of Node IPs is cloud-provider's responsibility and I will follow up this issue in cloud-provider, Node side.
I think your suggestion is good and will update the PR

File: pkg/kubelet/kubelet_pods.go
Comment (aojea) [inline pkg/kubelet/kubelet_pods.go:1640]: we have already some utils 
```suggestion
					if err, ok utilnet.IsDualStackIPStrings(hostIPs) ; err == nil && ok {
```

File: pkg/kubelet/kubelet_pods.go
Comment (aojea) [inline pkg/kubelet/kubelet_pods.go:1645]: I think we should give some clue to the user, we know the cause that is that the node-ip flag and the order returned by the ccm doesn't match, I think that something like this will make it
```suggestion
							"Inconsistent addresses order received from the cloud-provider, IP family of the second host IP %q is the same as the first one configured in the kubelet %q",
							hostIPs[1].String(), s.PodIPs[0].IP)
```

File: pkg/kubelet/kubelet_pods.go
Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go]: Thank you!. I think `err == nil && ok` should not be handled together. For `e2e != nil `, I use a different warning message about invalid host IP.

File: pkg/kubelet/kubelet_pods.go
Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go]: @aojea Hello! Do you expect other changes from my side?

File: pkg/kubelet/kubelet_pods.go
Comment (danwinship) [inline pkg/kubelet/kubelet_pods.go:1646]: The wording here is not great in terms of helping the admin understand what to do.

Also, if we are going to notice the "you requested IPv4 but the cloud provider wants dual-stack IPv6-primary" case, then we should really also notice the "you requested IPv4 but the cloud provider wants _single-stack_ IPv6" case too. (It is likely that the original reporter was also running into that case but just didn't notice yet because it doesn't cause any explicit errors. But it may cause problems later with Services.)

So, starting at line 1627:
- if `utilnet.IPFamilyOfString(s.HostIP) != utilnet.IPFamilyOf(hostIPs[0])` then we need to log a warning:
  - if `s.HostIP` is IPv4 then the problem is that kubelet auto-detected an IPv4 node IP, but the cloud provider wants this to be a single-stack IPv6 or dual-stack IPv6-primary node, and so the fix is for the admin to either (a) pass `--node-ip ::` to kubelet to tell it to prefer IPv6 like the cloud provider wants, (b) pass `--node-ip <EXPLICIT-IPV4-ADDRESS>` to request a single-stack IPv4 node with a specific IP. (Once `CloudDualStackNodeIPs` is beta we can also suggest (c) pass `--node-ip <EXPLICIT-IPV4-ADDRESS>,<EXPLICIT-IPV6-ADDRESS>`.)
  - if `s.HostIP` is IPv6 then the problem is that the admin passed `--node-ip ::` to request an IPv6 node IP, but the cloud provider wants this to be a single-stack IPv4 or dual-stack IPv4-primary node, and so the fix is for the admin to either (a) _not_ pass `--node-ip ::` to kubelet and let it default to IPv4-primary like the cloud provider wants, (b)/(c) pass an explicit IPv6 IP as above
- regardless of whether it logged a warning, it should do `s.HostIP = hostIPs[0].String()`, as it does now
- if the pod is host-network then:
  - it should set `s.PodIPs[0]` to `hostIPs[0]` if `s.PodIPs` is unset, as it does now
  - it should set `s.PodIPs[1]` to `hostIPs[1]` if `len(hostIPs) == 2 && len(s.PodIPs) == 1` _and_ `utilnet.IPFamilyOfString(s.PodIPs[0].IP) != utilnet.IPFamilyOf(hostIPs[1])`
    - (the family mismatch can only happen if we have already logged an error above, so we don't need to warn again here)

**Review by danwinship — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kubelet_pods.go
Comment (aojea) [reply pkg/kubelet/kubelet_pods.go]: sorry, went on vacation, Dan made a very detailed comment about some other edge cases that we may document and warn

**Comment by aojea** [comment]
I think is ok, but defer to Dan, he is driving all this node-ip effort and the person more familiar with it
/assign @danwinship @thockin

**Comment by lzhecheng** [comment]
@danwinship hello, do you have any other comments? Thank you.

File: pkg/kubelet/kubelet_pods.go
Comment (danwinship) [inline pkg/kubelet/kubelet_pods.go:1632]: I think

    if utilnet.IPFamilyOfString(s.HostIP) != utilnet.IPFamilyOf(hostIPs[0]) {

is clearer: the error case can't actually happen anyway, and we're comparing exactly two IPs, not scanning a possibly-long list of IPs.

File: pkg/kubelet/kubelet_pods.go
Comment (danwinship) [inline pkg/kubelet/kubelet_pods.go:1634]: if utilnet.IsIPv4(hostIPs[0]) {

actually, wait, this is backwards: `hostIPs` is what the cloud provider chose, `s.HostIP` is what kubelet detected.

File: pkg/kubelet/kubelet_pods.go
Comment (danwinship) [inline pkg/kubelet/kubelet_pods.go:1638]: The information that is in the comment needs to be in the Event text instead to make it understandable (after which you no longer need the comment).

``"Kubelet auto-detected an IPv4 node IP, but the cloud provider selected an IPv6 node IP; pass an explicit `--node-ip` to kubelet to fix this."``

File: pkg/kubelet/kubelet_pods.go
Comment (danwinship) [inline pkg/kubelet/kubelet_pods.go:1643]: likewise. Also, the fix here is to pass an explicit _IPv6_ address, not IPv4.

actually, we don't really even need separate errors for the v4 and v6 case. Just replace this all (1627-1646) with:

```
if s.HostIP != "" && utilnet.IPFamilyOfString(s.HostIP) != utilnet.IPFamilyOf(hostIPs[0]) {
        kl.recorder.Eventf(pod, v1.EventTypeWarning, "HostIPsIPFamilyMismatch",
                "Kubelet detected an IPv%s node IP (%s), but the cloud provider selected an IPv%s node IP (%s); pass an explicit `--node-ip` to kubelet to fix this.",
                utilnet.IPFamilyOfString(s.HostIP), s.HostIP, utilnet.IPFamilyOf(hostIPs[0])), hostIPs[0].String())
}
```

File: pkg/kubelet/kubelet_pods.go
Comment (danwinship) [inline pkg/kubelet/kubelet_pods.go:1658]: again, I think using `IPFamilyOf != IPFamilyOf` would be clearer here, and get rid of the unnecessary error check

File: pkg/kubelet/kubelet_pods.go
Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go]: I also prefer clearer code. Updated.

File: pkg/kubelet/kubelet_pods.go
Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go]: done

File: pkg/kubelet/kubelet_pods.go
Comment (lzhecheng) [reply pkg/kubelet/kubelet_pods.go]: done

**Comment by danwinship** [comment]
/lgtm
but needs a node approver

**Comment by lzhecheng** [comment]
/assign @Random-Liu 
@Random-Liu @thockin could you please help review the PR? It has lgtm now.

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**
